### PR TITLE
otel: setup otel for computed and storaged

### PR DESCRIPTION
**Pull request by guswynn**
This pr allows computed and storaged to be setup to send traces to an otel-collector. They use the exact same env vars, so that all services can be configured with a single endpoint env var, easily! Other than that, this is mostly code-movement and some copying.

### Motivation

  * This PR adds a known-desirable feature.

opentel support for all our services.

### Tips for reviewer

This will conflict with #12483 and Ill have to make sure they compose well

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
image

(I also tested a fake testing span for computed and it showed up, but normal traffic doesn't show up right now

### Release notes

None

### CODE REVIEW

**Comment by guswynn** [comment]
@benesch note that, now that this is rebased, computed and storaged dont produce any traces `debug` level and above, but I tested that they correctly can send traces locally

**Comment by benesch** [comment]
Mostly looks good, but the env vars should be consistent with the other env vars for each process (e.g., `COMPUTED_OPENTELEMETRY_HEADER` rather than `MZ_OPENTELEMETRY_HEADER`). To preserve the behavior that

```
bin/materialized -- --opentelemetry-endpoint=http://localhost:4317
```

enables tracing for all services, we should plumb the value of `opentelemetry_endpoint` and `opentelemetry_headers` _through_ Materialize's orchestrator and into the command-line arguments of `storaged` and `computed` services. This'll be a bit annoying, but it's what we need to do to make Kubernetes work anyway, since the computed/storaged services in Kubernetes don't share the environment variables of the controlling `materialized` process!

**Comment by guswynn** [comment]
@benesch should these args be passed through both the k8s and process orchestrators?

**Comment by benesch** [comment]
Yessir!

**Comment by guswynn** [comment]
@benesch new commit pushes args through the orchestrator. I probably clone more often than I have to, but they are once-per-service

Also, I don't pass through the `LOG_FILTER`, because `Targets` can't be turned back into a string, but I also feel that we should mirror the normal `LOG_FILTER` and make it per-service configurable!

**Comment by benesch** [comment]
Changes LGTM, modulo some docstrings that need to be written!

> Also, I don't pass through the `LOG_FILTER`, because `Targets` can't be turned back into a string, but I also feel that we should mirror the normal `LOG_FILTER` and make it per-service configurable!

I agree!

**Comment by benesch** [comment]
I took the liberty of adding the doc strings and rebasing... and then I started shaving a very hairy flock of yaks:

  * Enabling Tokio console for `materialized` now flows to all created services, with different ports assigned for each process.
  * Parsing of the tracing CLI arguments is extracted into a shared crate/struct so it doesn't need to be duplicated in every crate.
  * Injecting the correct tracing CLI arguments is handled by a new `TracingOrchestrator` that wraps another orchestrator. This is necessary prep work for a forthcoming PR that will push storage pod creation into the storage controller, where it otherwise wouldn't have access to the tracing configuration.
  * Filters for logging/otel now get plumbed through—I think this is actually the better default. If you're trying to e.g. debug librdkafka, you definitely want `--log-filter=librdkafka=debug` to flow through to `storaged`. If we ever do need per-service filters, we can introduce some `--log-filter-override=SERVICE=filter` flag, but didn't seem worth it for now!

I tested with Jaeger and the Tokio console locally and it all seemed to work great. Marking for automerge since it's crunch time and I want to get this in to unblock the pod-per-source work. But @guswynn definitely let me know if you have any post-merge feedback or just send PRs my way! And TYSM for getting this all wired up! It's so cool to see these distributed traces start to come together in Jaeger.

File: src/compute/src/bin/computed.rs [hunk 1]
```diff
@@ -90,22 +90,8 @@ struct Args {
...
-        value_name = "FILTER",
-        default_value = "info"
-    )]
-    log_filter: Targets,
-
-    /// Add the process name to the tracing logs
-    #[clap(long, hide = true)]
-    log_process_name: bool,
+    #[clap(flatten)]
+    tracing: TracingCliArgs,
```
Comment (guswynn) [inline src/compute/src/bin/computed.rs:94]: nice!

File: src/orchestrator-tracing/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    pub tokio_console_retention: Duration,
+}
+
+impl Default for TracingCliArgs {
+    fn default() -> TracingCliArgs {
+        let matches = TracingCliArgs::command().get_matches_from::<_, OsString>([]);
+        TracingCliArgs::from_arg_matches(&matches)
+            .expect("no arguments produce valid TracingCliArgs")
+    }
+}
```
Comment (guswynn) [inline src/orchestrator-tracing/src/lib.rs:158]: why do we need this impl?

File: src/orchestrator-tracing/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+impl FromStr for SerializableTargets {
+    type Err = tracing_subscriber::filter::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(SerializableTargets {
+            inner: s.parse()?,
+            raw: s.into(),
+        })
+    }
+}
```
Comment (guswynn) [inline src/orchestrator-tracing/src/lib.rs:325]: hell ya

**Review by guswynn — COMMENTED** [review]
this is a doozy to understand and review, but looks good!

**Pull request merged**